contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpHandler.java

```diff
@@ -0,0 +1,213 @@
+import io.opencensus.trace.Tracer;
+import io.opencensus.trace.propagation.TextFormat;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nullable;
+
+/**
+ * A helper class to handle instrumentation in HTTP client/server.
+ *
+ * <p>This class provides basic operations, such as recording {@link MessageEvent} and handling
+ * errors.
+ *
+ * @since 0.13
+ */
+public abstract class HttpHandler<Q, P> {
+
+  /** The Open Census tracing component. */
+  Tracer tracer;
+
+  /** The {@link TextFormat} used in HTTP propagation. */
+  TextFormat textFormat;
+
+  /** The {@link HttpExtractor} used to extract information from request/response. */
+  HttpExtractor<Q, P> extractor;
+
+  private AtomicLong sentMessageEventId = new AtomicLong(0);
+  private AtomicLong recvMessageEventId = new AtomicLong(0);
+
+  /** Package-protected constructor to allow access from subclasses only. */
+  HttpHandler(Tracer tracer, TextFormat textFormat, HttpExtractor<Q, P> extractor) {
+    checkNotNull(tracer, "tracer");
+    checkNotNull(textFormat, "textFormat");
+    checkNotNull(extractor, "extractor");
+    this.tracer = tracer;
+    this.textFormat = textFormat;
+    this.extractor = extractor;
+  }
+
+  /**
+   * Returns the tracing component.
+   *
+   * @return the tracing component.
+   * @since 0.13
+   */
+  public Tracer getTracer() {
+    return this.tracer;
+  }
+
+  /**
+   * Returns the {@link TextFormat} used in HTTP propagation.
+   *
+   * @return the {@code TextFormat} used in HTTP propagation.
+   * @since 0.13
+   */
+  public TextFormat getTextFormat() {
+    return this.textFormat;
+  }
+
+  /**
+   * Returns the {@link HttpExtractor} used to extract information from the request/response.
+   *
+   * @return the {@code HttpExtractor} used to extract information from the request/response.
+   * @since 0.13
+   */
+  public HttpExtractor<Q, P> getExtractor() {
+    return this.extractor;
+  }
+
+  /**
+   * Returns a span name determined by the given parameters.
+   *
+   * @param request the request entity.
+   * @param method the qualified name of the method which handles the request sending procedure,
+   *     which will be used when constructing the span name. Although user can obtain such
+   *     information from the request itself, this parameter provides a more convenient way to
+   *     specify it directly.
+   * @return a span name determined by the given parameters.
+   * @since 0.13
+   */
+  public abstract String getSpanName(Q request, String method);
+
+  /**
+   * Generate a sequential id for sent {@link MessageEvent}.
+   *
+   * @return next id for sent {@code MessageEvent}.
+   * @since 0.13
+   */
+  public long nextSentMessageEventId() {
+    return sentMessageEventId.getAndIncrement();
+  }
+
+  /**
+   * Generate a sequential id for received {@link MessageEvent}.
+   *
+   * @return next id for received {@code MessageEvent}.
+   * @since 0.13
+   */
+  public long nextRecvMessageEventId() {
```

Comment:
Why public? please make public only APIs that user must call. Also consider to make them final if you don't expect override.

Reply:
It is public to be used by user to when recording a new message event. We could hide them if we expose something like `recordSentMessageEvent`/`recordRecvMessageEvent` which automatically specify id.
It is not final because I would like user to extend them if they need something other than sequential increment behaviors.

I know that composition is better than inheritance but here this case is too subtle, so I didn't create another class to hold this API.